Add BaseLanguageHandling option to Datagen

sffc opened the pull request:
See https://github.com/unicode-org/icu4x/issues/58

I implemented the new option as `BaseLanguageHandling::Retain` or `BaseLanguageHandling::Strip`.

sffc commented on provider/datagen/src/driver.rs:
```diff
@@ -370,23 +385,32 @@ impl DatagenDriver {
                         .iter()
                         .try_for_each(|(locale, (payload, _duration))| {
                             let mut iter = fallbacker_with_config.fallback_for(locale.clone());
+                            let mut first = true;
```
The effect of `first` is that it prevents non-base locales such as `tlh-001` from being retained. I'm not sure if this is the behavior that we desire; it only comes up when an unsupported locale is requested and that locale is not a base language. I lean toward removing the check on `first`, which would cause `tlh-001` to be retained.

jedel1043 replied:
I think it's fine to remove `first`, but we should probably print a warning alerting the user that the provider locale is not supported.

sffc replied:
Yep. Today I'm thinking that we probably should not include _either_ `tlh-001` _or_ `arc` since there is no CLDR data for them, and therefore they are not really "supported". I'll work on that tweak.

robertbastian commented on provider/datagen/src/driver.rs:
```diff
@@ -109,6 +112,16 @@ impl DatagenDriver {
+    /// Sets the base language handling in runtime fallback mode.
+    ///
+    /// For more details, see [`BaseLanguageHandling`].
+    pub fn with_base_language_handling(self, base_language_handling: BaseLanguageHandling) -> Self {
```
issue: this option is only valid for a specific fallback mode. I'd prefer modelling it with a new fallback mode

jedel1043 replied:
I think it applies for both `Runtime` and `RuntimeManual`, in addition to applying to `PreferredForExporter` in certain cases; an user who wants to enable resolvedLocale queries but doesn't want to pick a specific fallback type. This would mean we would need to add 3 more fallback variants + any other variant that is added in the future that is similar to `Runtime`.

sffc replied:
Good point about `PreferredForExporter`. With that option, the exporter can state its preferred fallback mode, but that question is completely separate from whether to include the base languages.

robertbastian replied:
An exporter does not state its preferred fallback mode, it states `supports_built_in_fallback`. Datagen translates this to an actual mode, which we are free to change: `true` can be `RuntimeThin`, and `false` `HybridThin`, as you are currently making this the default already.

robertbastian commented on provider/datagen/src/lib.rs:
```diff
@@ -225,6 +225,9 @@ pub enum FallbackMode {
     /// they contain the same value, only "en" will be included, since "en-US" falls back to
     /// "en" at runtime.
     ///
+    /// By default, locales that inherit directly from "und" will be retained even if their data
+    /// is equal to "und". To override this behavior, see [`BaseLanguageHandling`].
+    ///
```
issue: I'm not convinced this is a good default. This increases baked data size, but supported locale queries are not a universal use case.

sffc replied:
Happy to discuss the default. Given that it's needed for 402 and we've seen enough examples of use cases where people may want to do these queries out of the box, and the data size increase is fairly small (no regional variants), I lean toward including these by default. It's easy for power users to get the extra 5% data size benefit if they know what they are doing.

Manishearth replied:
slightly lean towards the default  of Strip. I think either way we will need careful documentation of supported locale queries and how they can work. We should do that documentation and clearly document the defaults.

robertbastian commented on provider/datagen/src/lib.rs:
```diff
@@ -276,6 +279,28 @@ pub enum FallbackMode {
+/// A _base language_ is a locale that inherits directly from `und`. For example, `en` and
+/// `zh-Hant` inherit directly from `und`.
+///
+/// Removing base languages whose data equals `und` may reduce data size, but it may cause
+/// supported-locale queries to fail.
```
```suggestion
/// Removing base languages whose data equals `und` may reduce data size, but makes it impossible to determine "supported" locales.
```

sffc replied:
Reworded

sffc commented:
Current:

- Runtime
- RuntimeManual
- Hybrid
- Preresolved

Proposed to add:

- RuntimeWithBaseLanguages
- RuntimeManualWithBaseLanguages

---

- @sffc - How to we turn off regional variants?
- @robertbastian - Maybe something like `en-XX` to not include regional variants, this woulud be more granular because you could do `de,en-ZZ` for all `de` variants and no `en` variants
- @sffc - I think it should be a variant, because it can also apply to regional variants like `en-001`. So you may want `en-001-nochilds`, which includes `en` and `en-001` but not `en-GB`.

---

- @robertbastian - Maybe we want something like

```rust
Deduplicated / Runtime2 {
  use_internal_fallback: bool,
  include_base_languages: bool,
}
```

Now:

```rust
#[non_exhaustive]
pub enum FallbackMode {
    #[default]
    DefaultForProvider,
    // same as Deduplicated { RuntimeFallbackLocation::Internal, BaseLanguageHandling::Strip }
    Runtime,
    // same as Deduplicated { RuntimeFallbackLocation::External, BaseLanguageHandling::Strip }
    RuntimeManual,
    Hybrid,
    Preresolved,
    Deduplicated(Deduplicated),
}
```

2.0:

```rust
#[non_exhaustive]
pub enum LocaleExpansionMode {
    Deduplicated(Deduplicated),
    Exhaustive(Exhaustive),
    Preresolved(Preresolved),
}

#[non_exhaustive]
pub enum RuntimeFallbackLocation {
    Internal,
    External,
}

#[non_exhaustive]
pub enum BaseLanguageHandling {
    #[default]
    Retain,
    Strip,
}

#[non_exhaustive]
#[derive(Default)]
pub struct Deduplicated {
  runtime_fallback_location: Option<RuntimeFallbackLocation>,
  base_language_handling: BaseLanguageHandling,
}

// resolve this against the exporter like
foo.runtime_fallback_location.unwrap_or_else(|| 
    if exporter.supports_runtime_fallback() {
        RuntimeFallbackLocation::Internal
    } else { 
        RuntimeFallbackLocation::External 
    }
)

#[non_exhaustive]
pub struct Exhaustive {}

#[non_exhaustive]
pub struct Preresolved {}
```

LGTM: @Manishearth @sffc @robertbastian

sffc commented:
@robertbastian What do you want the CLI to look like in 1.5 and in 2.0?

robertbastian commented:
I guess for now we flatten the enum like `runtime-manual-retain‘ and ‘runtime-strip‘?

sffc commented:
If we are going to end up with a `base_language_handling` option in the `Deduplicated` enum, then maybe it makes sense that it should be its own flag on the CLI.

sffc commented:
We have a lot of flags that are only used if some other flag is set. This would be the same.

sffc commented:
@robertbastian I would like to merge this as-is because:

1. The final 2.0 CLI option should indeed be named `--base-language-handling` because that is how we handle all other cases where flags are not used depending on the value of other flags. If you really want I could make `--base-language-handling` return an error if set with the wrong value for `--fallback`.
2. In the API, we are changing it anyway in 2.0 as discussed above. The change proposed is the smallest change to the API. I don't want to implement only half of the 2.0 changes now. I would rather we, in another PR, implement LocaleExpansionMode as designed above. We could even do that in 1.5 and deprecate FallbackMode. In that PR, we could delete any unreleased APIs such as this one that are covered by LocaleExpansionMode.

jedel1043 commented:
With my engine maintainer hat on, I have to agree with @sffc. Boa will require all engine embedders to use this feature, so from a docs perspective it's easier to guide our users to activate a simple flag than to explain to them which fallback variants are supported and how to activate them.

robertbastian reviewed (approved):
I still consider this additional flag tech debt.

Manishearth commented on provider/datagen/src/bin/datagen/args.rs:
```diff
@@ -232,8 +239,17 @@ pub struct Cli {
+    #[arg(long, value_enum, default_value_t = BaseLanguageHandling::Auto)]
+    #[arg(
+        help = "--fallback=runtime[-manual] only: Whether to retain or strip base languages. \
+                For details, read the Rust docs for BaseLanguageHandling."
```
nit: let's call the local one BaseLanguageHandlingArg, to avoid confusion when looking at the code. The one above has no docs.

Alternatively give it docs saying to look at the type in the config module.

robertbastian reviewed (changes requested):
There's a lot of discussion going on in #4629 which I'd like to resolve before merging anything.

sffc commented:
Need to re-do this on top of #4710

jedel1043 commented:
Probably super seeded by #4836.

sffc commented:
> Probably super seeded by #4836.

Things this PR does that #4836 didn't do:

- Added to CHANGELOG.md
- Added "tlh-001" to the options test (to test an unsupported non-base locale)
